[Windows] Support original_dst/sidecar

davinci26 opened the pull request:
Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>
Commit Message:

Fixes #13955

Additional Description:

The architecture on Windows is the following:

1. When the original_dst filter is used on Windows, we use `SO_ORIGINAL_DST` to query for the original address.
2. If the traffic direction is `Outbound` and address has been restored. Then query for the WFP redirect records (think about this as the firewall rules) and apply the redirect records to the outbound connection.


### Pending work

### Code & Testing
- [x] Validate case where traffic is ipv6
- [x] Make a cohesive story of the `#ifdefs` to ensure a smooth experience between older and newer platforms
- [x] Add unit/integration tests
- [x] Add validation for `original_dst` listeners to also use `traffic_direction`

### Documentation
- [x] Document design choices regarding `traffic_direction`
- [x] Document platform version support
- [x] Document how to build it with the newer Windows SDK

Risk Level: Low (Windows only, it should be a null-op for Unix)
Testing: Unit + Manual
Docs Changes: Document the usage of inbound traffic property
Release Notes:  Adds support for the original.dst filter on Windows
Platform Specific Features: Windows Parity

davinci26 commented:
@ggreenway some points that I am looking for some feedback on:

1. is `stream_info` appropriate info to store the redirect records
2. do I break any existing connection handling scenario?
3. Should the logic for getting the traffic direction be moved  to the filter level?

davinci26 commented:
I have pending work to test this ( I have been testing this manually so far) but I am waiting for some feedback on the high level structure direction of the changes and then I can move into that

ggreenway commented:
Given that this data isn't present most of the time (only when using original-dst, and only on windows), it might come out cleaner to set something in the `filterState()` on the stream info instead of adding it at the top-level on `StreamInfo`.

ggreenway commented:
Does the original dst stuff only work for direction == outbound on linux? A quick skim of the source makes it appear that `direction` is only used for draining-related behavior. So could you just omit the check for direction?

davinci26 commented:
> Given that this data isn't present most of the time (only when using original-dst, and only on windows), it might come out cleaner to set something in the filterState() on the stream info instead of adding it at the top-level on StreamInfo.

Seems reasonable. I will look into it.

> Does the original dst stuff only work for direction == outbound on linux?

Why do you think so? My understanding is that Istio configures a virtual inbound listener with the `original_dst` property to capture incoming traffic.

> A quick skim of the source makes it appear that direction is only used for draining-related behavior.

Quite likely. AFAIK, I am the only one that uses the `traffic_direction` property in that way. The issue is that on Windows we need to distinguish between inbound and outbound traffic to apply the redirect records. This was the best way I could find to distinguish them. From inspections in Istio codebase,  it always applies that property.

ggreenway commented:
> > Does the original dst stuff only work for direction == outbound on linux?
> 
> Why do you think so? My understanding is that Istio configures a virtual inbound listener with the `original_dst` property to capture incoming traffic.

Maybe I'm misunderstanding (or I misread your change). It looked to me like the change you're making only applies to listeners with direction == OUTBOUND. Can you explain the intention of that? 

> > A quick skim of the source makes it appear that direction is only used for draining-related behavior.
> 
> Quite likely. AFAIK, I am the only one that uses the `traffic_direction` property in that way. The issue is that on Windows we need to distinguish between inbound and outbound traffic to apply the redirect records. This was the best way I could find to distinguish them. From inspections in Istio codebase, it always applies that property.

Can you elaborate on that? I just want to understand the flow of how this works on windows, and the constraints.

davinci26 commented:
> It looked to me like the change you're making only applies to listeners with direction == OUTBOUND. Can you explain the intention of that?

Sorry, I realized something must have not been clear just by reading the code. With the newer Windows SDK the definition `SO_ORIGINAL_DST` is **now** defined on Windows. So the existing code for inbound traffic traffic works as it is. This is why there are no changes for the inbound direction.

We have two cases for traffic redirection:

1. **Inbound traffic** which is handled by a Virtual Inbound Listener. In this case the Envoy code work as is, assuming a version of Windows that supports `SO_ORIGINAL_DST` socket option.
2. **Outbound traffic** which is handled by a Virtual Outbound Listener. In this case Windows and UNIX divert (hence the changes). For outbound traffic, on Windows, we need to respect the existing firewall rules of the container. 

Imagine the following scenario, which is only applicable for the outbound traffic direction:

Players: *container*, Sidecar Container, google.com (upstream)

When *container* makes a request to google.com the traffic (outbound) gets intercepted by the sidecar container. This means that we have (i) a socket between *container*(downstream) and sidecar and (ii) another socket between sidecar and google.com (upstream).

In this scenario the platform admin might have set firewall rules blocking traffic from container 1 to google.com. The sidecar should respect these rules when proxying the connection. This means that we should query for the redirect records of the downstream socket and re apply these records to the upstream socket.

The bulk code changes in this PR are to support the logic described above.

> Can you elaborate on that? I just want to understand the flow of how this works on windows, and the constraints.

The above lead us to a subtle problem. Since Envoy needs to do "extra" work on outbound connections, how can it distinguish between the two cases?

Here I considered 3 approaches:

1. Rely on the IP of the address and the destination. I think this is a fragile approach with many edge cases like handling traffic within the same POD.
2. Try to apply the redirect records in both inbound and outbound traffic and accept that this will fail 50% of the time. This is a bit awkward in terms of the user experience. Because if something goes wrong and the outbound direction does not work then we would silently fail thinking the connection is inbound.
3. Use the listener `traffic_direction` property and rely on service mesh frameworks setting this property to the listeners. I have verified that Istio already does that.

IMHO approach (3) is the best but I am always open to suggestions.

ggreenway commented:
Thanks, that makes more sense now.

You'll need to document somewhere that this flow won't work on windows unless the traffic direction is configured properly. I don't think the traffic direction configuration affects this functionality on linux. But I agree that this is a reasonable use of that configuration, in order to get what you need to make it work on windows.

ggreenway commented:
@envoyproxy/api-shepherds care to chime in? This wouldn't change any .protos, but it is new use of existing config.

davinci26 commented:
> You'll need to document somewhere that this flow won't work on windows unless the traffic direction is configured properly.

+1, for sure

davinci26 commented:
I incorporated the first comment and added the `redirect_records` into a `filter_state` instead of `stream_info`

htuch commented:
@ggreenway @davinci26 can you provide a TL;DR from a user/API perspective? I.e. if I configure original_dst in a management server and I speak to a Linux and Windows Envoy, what differences would I expect for traffic? What assumptions about the environment?

davinci26 commented:
> if I configure original_dst in a management server and I speak to a Linux and Windows Envoy, what differences would I expect for traffic?
 
The experience should the similar. In the sense that the traffic get's intercepted and redirected from the application container to the sidecar container. Note that this works well for container traffic redirection.

Also we the platform only supports redirection for tcp and not udp.
 
 > What assumptions about the environment?
 
 1. A version of Windows that supports traffic redirection (support `SOL_IP` and `SO_ORIGINAL_DST`).
 2. Traffic redirection has been setup for the environment. In Windows we don't have `iptables` so the way to set this up is by modifying the `HNS` policies.
 3. Set the `traffic_direction` property into the listener.

htuch commented:
This seems like a reasonable implementation without API semantic divergence, @ggreenway did you have some concern around this from the API side?

ggreenway commented:
@htuch My only concern is that this setup would work correctly on linux without setting the traffic direction, but the direction is required on windows. But given that windows requires this additional information, I think using the existing direction is better than creating a new field for this. I just wanted another opinion on it from an API PoV.

htuch commented:
@ggreenway yeah, I agree, I think as long as we can sensibly configure both Linux/Windows with the traffic direction field set for consistent behavior, then we're all good. The idea we're aiming for is that a management server can be client neutral, i.e. not condition its behavior on the OS. In this case, there's a reasonable way to do this I think.

ggreenway commented:
@davinci26 One other thing to try/consider: could the code for getting the redirect record be put in the original_dst listener filter, as `dynamicMetadata` (which can be set on the listener filter callbacks)?

I think the code in router/tcp_proxy and cluster makes sense, for handing off that information at the various points.

@lizan Do you have any thoughts on the approach to this?

davinci26 commented:
> One other thing to try/consider: could the code for getting the redirect record be put in the original_dst listener filter, as dynamicMetadata (which can be set on the listener filter callbacks)?

I think that we could do that but we would need to know `traffic_direction` property of the filter listener on the filter. I could not find any precedent of filters being aware of the configuration of the listener.

Do we have any examples were we pass additional info like the above to the filters? If we can do it at the filter level the info could be store as part of the restored `local_address` of the `socket` and then move to `filter_state` by the connection handler.

davinci26 commented:
@ggreenway when you get some time, this should be ready for another review.

@rmiller14 this should be a safe commit for internal testing. Feel free to hop in for a review as well.

davinci26 commented on source/common/network/socket_option_factory.cc:
```diff
@@ -52,6 +53,15 @@ std::unique_ptr<Socket::Options> SocketOptionFactory::buildIpTransparentOptions(
+  std::unique_ptr<Socket::Options> options = std::make_unique<Socket::Options>();
+  options->push_back(std::make_shared<Network::IoctlSocketOptionImpl>(
+      envoy::config::core::v3::SocketOption::STATE_PREBIND, ENVOY_SOCKET_REDIRECT_RECORDS,
+      (void*)redirectRecords.buf_ptr_, redirectRecords.buf_size_, nullptr, 0));
```
use reinterpreted cast here instead of c style cast

davinci26 replied:
fixed

davinci26 commented on source/common/network/socket_option_factory.cc:
```diff
@@ -52,6 +53,15 @@ std::unique_ptr<Socket::Options> SocketOptionFactory::buildIpTransparentOptions(
 }
 
+std::unique_ptr<Socket::Options>
+SocketOptionFactory::buildWFPRedirectRecordsOptions(const EnvoyRedirectRecords& redirectRecords) {
```
follow convention for variable names

davinci26 replied:
fixed

davinci26 commented on source/common/network/socket_option_factory.h:
```diff
@@ -25,6 +25,8 @@ class SocketOptionFactory : Logger::Loggable<Logger::Id::connection> {
   static std::unique_ptr<Socket::Options> buildIpFreebindOptions();
   static std::unique_ptr<Socket::Options> buildIpTransparentOptions();
+  static std::unique_ptr<Socket::Options>
+  buildWFPRedirectRecordsOptions(const EnvoyRedirectRecords& redirectRecords);
```
same as above

davinci26 replied:
fixed

ggreenway commented on source/common/network/ioctl_socket_option_impl.h:
```diff
@@ -0,0 +1,55 @@
+  const envoy::config::core::v3::SocketOption::SocketState in_state_;
+  const Network::SocketOptionName optname_;
+
+  void* inBuffer_;
```
It's hard to figure out what data this is pointing to, and the relative lifespan of this object vs the thing that owns the buffer.

davinci26 replied:
> It's hard to figure out what data this is pointing to

This an opaque data structure since `WSAIoctl` accepts different byte arrays depending on the control code, so it is hard to make it more intuitive. I opted to makes this option class as generic as possible for future usecase/calls to `WSAIoctl`.

I agree that the ownership semantics here are not clear. I will look into improving them

davinci26 replied:
Do you think its a better idea to have this class more specific to this feature `Win32RedirectRecordsOptionImpl`? This would make it easier to understand the semantics? This is less abstract but we can make the abstraction forward if we need more `Ioctl` socket options

davinci26 replied:
fixed in 063cc63b664dde3d008b642bc53c0bcec7399d10

ggreenway commented on source/common/network/ioctl_socket_option_impl.cc:
```diff
@@ -0,0 +1,57 @@
+IoctlSocketOptionImpl::getOptionDetails(const Socket&,
+                                        envoy::config::core::v3::SocketOption::SocketState) const {
+
+  std::string in_buffer_bstr(reinterpret_cast<const char*>(inBuffer_), inBuffer_size_);
```
naming: `_bstr`: what does that mean?

ggreenway replied:
absl::string_view would remove a copy

davinci26 replied:
> naming: _bstr: what does that mean?

Binary string. I took the name from other places in the codebase that did a similar cast and kept the suffix for consistency. No strong opinion though, I can change it

davinci26 replied:
> absl::string_view would remove a copy

This is a bit tricky because we need to concatenate the strings and this would require a copy since `string_view ` does not support either `insert` or `operator +`. I will think about it more

ggreenway commented on source/common/network/ioctl_socket_option_impl.h:
```diff
@@ -0,0 +1,55 @@
+namespace Envoy {
+namespace Network {
+
+#define IOCTL_LEVEL 1
```
Where did `1` come from? Is it windows specific?

davinci26 replied:
To comply with the `socket_option` interface in `#define ENVOY_MAKE_SOCKET_OPTION_NAME(level, option)` I need to set a level.

I can probably set it to `"IOCTL"` which will make it better for the reader

davinci26 replied:
I added a comment in 063cc63b664dde3d008b642bc53c0bcec7399d10 and renamed it to be look Envoy specific

ggreenway commented on source/server/connection_handler_impl.cc:
```diff
@@ -477,6 +478,35 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
   stream_info->setDownstreamSslConnection(transport_socket->ssl());
+
+  if constexpr (Network::Win32SupportsOriginalDestination()) {
+    ASSERT(socket->addressProvider().localAddressRestored() &&
```
Isn't `socket->addressProvider().localAddressRestored()` false for all non-original-dst listeners?

davinci26 replied:
You are correct, regardless I removed the assert in 583512a9750f17a04e7c1de4d0a428588fda08a1 since we will be guarding this in `listener_impl`.

ggreenway commented on source/server/connection_handler_impl.cc:
```diff
@@ -477,6 +478,35 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
+      stream_info->filterState()->setData(
+          Network::RedirectRecordsFilterState::key(),
+          std::make_unique<Network::RedirectRecordsFilterState>(redirect_records),
+          StreamInfo::FilterState::StateType::Mutable,
```
ReadOnly? I don't think this is ever changed on a connection, is it?

davinci26 replied:
I will try it out with `ReadOnly` althought I am not sure if it will work because the `IOCTL` api does not guarantee that it will not modify the buffer.

davinci26 replied:
Changed to readonly in 583512a9750f17a04e7c1de4d0a428588fda08a1

ggreenway commented on source/server/connection_handler_impl.cc:
```diff
@@ -502,7 +532,7 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
     LinkedList::moveIntoList(std::move(active_connection), active_connections.connections_);
   }
-}
+} // namespace Server
```
revert? I think clang-format probably did this when you didn't want it.

davinci26 replied:
clang-format doesnt play well with `#ifdefs` so it adds it. Now with the constexpr function it formats correctly

davinci26 replied:
fixed in 583512a9750f17a04e7c1de4d0a428588fda08a1

ggreenway commented on source/server/listener_impl.cc:
```diff
@@ -372,6 +372,32 @@ ListenerImpl::ListenerImpl(ListenerImpl& origin,
+                           "specifying the traffic_direction."
+                           "Configure the traffic_direction listener option");
+    } else {
+      ENVOY_LOG(warn,
```
I'm not sure I like the idea of different config being valid on only some versions of windows.I think the above validation should be done whether or not `Win32SupportsOriginalDestination()` is true. 

@htuch Any differing opinion on that? I'm also slightly uncomfortable with a config leaving direction unspecified, which will load and work on non-windows, and not be loadable on windows, but I think it's probably better than the alternative (loading a config we know does not work).

davinci26 replied:
I will refactor the if statement here to better reflect what the situation is.

davinci26 replied:
The code is now refactored in the following way:

```
if original_dst filter exists:
   if traffic_direction is unspecified:
      throw exception
   
   if windows versions does not support it:
      log warning
```

htuch replied:
IMHO we should throw if not supported; warnings are a silent failure and not particularly robust for most operators. 

Also, should this be done inside the `original_dst` filter factory config implementation? I think the listener is a strange place to do the validation.

davinci26 replied:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/71ca1453e586dbd51f07d08e5efa02563789c97d#diff-bf7c14c67a2a4f213ec7d90ee3a353c7c25381e11d5cec70a40c0093c007dff8R26 regarding where is validation is done.

I did not want to break compatibility this is why I opted for a warning. Given that currently if you specify the `original_dst` filter Envoy loads but does not work. I agree with @htuch that throwing an exception is the best way to fail fast.

@envoyproxy/windows-dev feel free to chime in.

davinci26 replied:
Also another thing is that on POSIX if `SOL_IP` is not defined then the user can still use the original destination filter even though it wont work. I am not sure how common this scenario is though

ggreenway reviewed (changes requested):
/wait

davinci26 commented on source/common/tcp_proxy/tcp_proxy.cc:
```diff
@@ -422,21 +424,24 @@ Network::FilterStatus Filter::initializeUpstreamConnection() {
-             .filterState()
-             ->hasData<Network::ProxyProtocolFilterState>(
-                 Network::ProxyProtocolFilterState::key())) {
-      read_callbacks_->connection().streamInfo().filterState()->setData(
```
I got confused and I thought it was a copy paste error on my side. I will revert

davinci26 replied:
reverted in 583512a9750f17a04e7c1de4d0a428588fda08a1

davinci26 commented on source/common/tcp_proxy/tcp_proxy.h:
```diff
@@ -377,6 +381,8 @@ class Filter : public Network::ReadFilter,
   RouteConstSharedPtr route_;
   Router::MetadataMatchCriteriaConstPtr metadata_match_criteria_;
   Network::TransportSocketOptionsSharedPtr transport_socket_options_;
+  Network::Socket::OptionsSharedPtr upstream_options_ =
```
I think `nullptr` is better, I will change it to that

davinci26 replied:
fixed in 583512a9750f17a04e7c1de4d0a428588fda08a1. Now it is allocated only when it is about to get used

davinci26 commented on include/envoy/network/io_handle.h:
```diff
@@ -26,6 +26,19 @@ using RawSliceArrays = absl::FixedArray<absl::FixedArray<Buffer::RawSlice>>;
 
 namespace Network {
 
+struct EnvoyRedirectRecords {
```
Fixed in 583512a9750f17a04e7c1de4d0a428588fda08a1

davinci26 commented on include/envoy/network/io_handle.h:
```diff
@@ -26,6 +26,19 @@ using RawSliceArrays = absl::FixedArray<absl::FixedArray<Buffer::RawSlice>>;
 namespace Network {
 
+struct EnvoyRedirectRecords {
+  void* buf_ptr_[256];
```
Makes sense, fixed in 583512a9750f17a04e7c1de4d0a428588fda08a1

davinci26 commented:
/retest

ggreenway reviewed (changes requested):
/wait

davinci26 commented on source/common/network/ioctl_socket_option_impl.cc:
```diff
@@ -35,20 +35,22 @@ bool IoctlSocketOptionImpl::setOption(
+  absl::string_view out_buffer_bstr(reinterpret_cast<const char*>(outBuffer_), outBuffer_size_);
   pushScalarToByteVector(
-      StringUtil::CaseInsensitiveHash()(std::string(in_buffer_bstr + out_buffer_bstr)), hash);
+      StringUtil::CaseInsensitiveHash()(std::string(in_buffer_bstr) + std::string(out_buffer_bstr)),
```
Ohhhhh I couldn't find this api online, I was looking for insert or operator +. It is not needed in the current implementation but thanks for sharing this. now I know this api exists

davinci26 commented on source/server/connection_handler_impl.cc:
```diff
@@ -480,18 +480,15 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
+      auto redirect_records = std::make_shared<Network::Win32RedirectRecords>();
+      auto status =
+          socket->genericIoctl(SIO_QUERY_WFP_CONNECTION_REDIRECT_RECORDS, NULL, 0,
+                               reinterpret_cast<uint8_t*>(redirect_records->buf_),
```
you are right, fixed in 063cc63b664dde3d008b642bc53c0bcec7399d10

davinci26 commented on include/envoy/api/os_sys_calls.h:
```diff
@@ -185,6 +185,14 @@ class OsSysCalls {
+  virtual SysCallIntResult genericIoctl(os_fd_t sockfd, unsigned long controlCode, void* InBuffer,
+                                        unsigned long InBufferLen, void* OutBuffer,
+                                        unsigned long OutBufferLen,
+                                        unsigned long& bytesReturned) PURE;
```
Just a personal preference for references over pointers but happy to make it consistent across the project. fixed in 063cc63b664dde3d008b642bc53c0bcec7399d10.

davinci26 commented on include/envoy/api/os_sys_calls.h:
```diff
@@ -185,6 +185,14 @@ class OsSysCalls {
+  /**
+   * @see MSDN WSAIoctl. Controls the mode of a socket.
+   */
+  virtual SysCallIntResult genericIoctl(os_fd_t sockfd, unsigned long controlCode, void* InBuffer,
```
> Should this be named win32Ioctl or similar?

No strong opinion here. I just did not want to add something that is `win32` specific into the API.

> Style: lowercase snake_case for parameter names.

fixed in 063cc63b664dde3d008b642bc53c0bcec7399d10

ggreenway replied:
The doc-comment references MSDN WSAIoctl. I think it should either be `ioctl` or `win32Ioctl`. I don't think `generic` adds any useful information.

davinci26 replied:
makes sense, I will rename it to  `win32Ioctl`

davinci26 commented on source/common/api/posix/os_sys_calls_impl.cc:
```diff
@@ -281,5 +281,10 @@ SysCallBoolResult OsSysCallsImpl::socketTcpInfo([[maybe_unused]] os_fd_t sockfd,
 
+SysCallIntResult OsSysCallsImpl::genericIoctl(os_fd_t, unsigned long, void*, unsigned long, void*,
+                                              unsigned long, unsigned long&) {
+  PANIC("genericIoctl not implemented on POSIX");
```
fixed in 063cc63b664dde3d008b642bc53c0bcec7399d10

davinci26 commented on source/common/network/redirect_records_filter_state.h:
```diff
@@ -0,0 +1,24 @@
+  static const std::string& key();
+
+private:
+  const std::shared_ptr<Network::Win32RedirectRecords> records_;
```
The more `const` the better, added in 063cc63b664dde3d008b642bc53c0bcec7399d10

sunjayBhatia reviewed (commented):
Overall summary of how this works and your changes to validate I understand correctly:
- an app container+Envoy sidecar share a network compartment and have HNS policies applied to do egress traffic redirection
- HNS uses WFP to do egress traffic redirection from the app container to the Envoy sidecar
- WFP modifies the redirect records on the Envoy listen socket being redirected to
- Envoy must copy those redirect records to the upstream socket it is connecting to
- Also, if the original destination filter is applied, Envoy will query the socket option to get the original address the downstream connection originated from (no difference between Windows and Linux)

Would be good to include maybe this link https://docs.microsoft.com/en-us/windows-hardware/drivers/network/sio-query-wfp-connection-redirect-records or https://docs.microsoft.com/en-us/windows-hardware/drivers/network/using-bind-or-connect-redirection to explain why the redirect records need to be queried from the downstream and set on the upstream socket

Changes for the original destination filter seem straightforward, makes sense to couple it with the changes to support redirection

davinci26 commented:
@sunjayBhatia yeah you got this all correct.

Will add a link to https://docs.microsoft.com/en-us/windows-hardware/drivers/network/using-bind-or-connect-redirection, this is good public documentation

davinci26 commented:
/retest

davinci26 commented:
@ggreenway I expect the CI to go green now, PTAL.

I have modified the listener `.proto` to add some more comments because they are used in documentation to communicate the new use for the `traffic_direction` property

ggreenway commented on source/common/network/win32_redirect_records_option_impl.h:
```diff
@@ -11,22 +11,24 @@
-      : in_state_(in_state), optname_(optname), inBuffer_(inBuffer), inBuffer_size_(inBufferSize),
-        outBuffer_(outBuffer), outBuffer_size_(outBufferSize) {}
+  Win32RedirectRecordsOptionImpl(envoy::config::core::v3::SocketOption::SocketState in_state,
+                                 Network::SocketOptionName optname,
```
I think `in_state` and `optname` can be hard-coded into the code for this class, instead of being parameters/member-variables.. Unless you have plans to use this class in another way.

davinci26 replied:
`optname ` is needed to check if the option `hasValue()` but the other is not needed. Fixed in 3437dc6078878c8282a466a53d09441d3b8437da

ggreenway replied:
I'm pretty sure you can use `ENVOY_SOCKET_REDIRECT_RECORDS` directly, as in `ENVOY_SOCKET_REDIRECT_RECORDS.hasValue()`. The macro evaluates to a `SocketOptionName`, which is either empty or has the correct value.

davinci26 replied:
Wouldn't that construct a new instance of `struct SocketOptionName` each time we call `hasValue()`?

ggreenway replied:
I think you could have a `static constexpr SocketOptionName` initialized from the macro in the file and use that everytime.

davinci26 replied:
I have some concerns with deleting `optname` because of the following:

1. The class is not trivially constructable. which means that this would violate envoy style guide,

> The Google C++ style guide points out that non-PoD static and global variables are forbidden. This includes types such as std::string. We encourage the use of the advice in the C++ FAQ on the static initialization fiasco for how to best handle this.

2. It breaks consistency will all other socket options.


That being said, I see what you are saying. Since all the socket option support info is known at compile time we don't need to initialize it every time we construct a socket option instance. I think this is a problem with how all the socket options behave right now.

davinci26 replied:
https://godbolt.org/z/jqTW7T

ggreenway replied:
I think it's ok to break consistency with other socket options, given that this one is kind of special (actually ioctl, not sockopt; uses special data). 

You could make one with `CONSTRUCT_ON_FIRST_USE`, or even just make a new instance on the stack everytime you need it; it's construction isn't a performance concern. With what you have now, you're creating a new instance everytime the object is created at the call to the constructor.

davinci26 replied:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/c2987c9223a969066c85373795ef7dec8ade79c9

ggreenway commented on source/common/network/win32_redirect_records_option_impl.h:
```diff
@@ -44,11 +46,7 @@ class IoctlSocketOptionImpl : public Socket::Option, Logger::Loggable<Logger::Id
-  unsigned long inBuffer_size_;
-  void* outBuffer_;
-  unsigned long outBuffer_size_;
+  Win32RedirectRecords redirect_records_;
```
Perfect. This is now much easier to reason about ownership/lifetime.

ggreenway commented on docs/root/configuration/listeners/listener_filters/original_dst_filter.rst:
```diff
@@ -3,13 +3,41 @@
+===============
+
+Original destination listener filter reads the SO_ORIGINAL_DST socket option set when a connection
+has been redirected by an HNS policy applied to a container endpoint. For this fiter to work the
```
Will the reader understand what `HNS` means from context?

ggreenway replied:
Spelling: fiter

davinci26 replied:
Doesn't hurt to add a link to make sure that it makes sense for everyone, added in 3437dc6078878c8282a466a53d09441d3b8437da

ggreenway reviewed (changes requested):
Sorry, ran out of time at the end of workday; I haven't finished going through all of this yet; I'll try to finish tomorrow morning. But there's what I have so far.

davinci26 commented on docs/root/configuration/listeners/listener_filters/original_dst_filter.rst:
```diff
@@ -3,13 +3,42 @@
 has been redirected by an iptables REDIRECT target, or by an iptables TPROXY target in combination
 with setting the listener's :ref:`transparent <envoy_v3_api_field_config.listener.v3.Listener.transparent>` option.
+
+Windows
```
@pravb and @elweb9858 do you think we should elaborate more in Envoy docs?

pravb replied:
This looks good to me. IMO sidecar redirection should be covered by a blog post which could note the differences between the platforms.

pravb commented on docs/root/configuration/listeners/listener_filters/original_dst_filter.rst:
```diff
@@ -3,13 +3,42 @@
+    At the time of writing (February 2021) the OS support for original destination is only available through the
+    `Windows insider program <https://insider.windows.com/en-us/for-developers>`_.
+    The feature will be fully supported in Windows Server 20H2
+    `Windows Server Release info <https://docs.microsoft.com/en-us/windows-server/get-started/windows-server-release-info>`_.
```
Not in 20H2. It will be in the next Windows Server release. We don't have a public name for it yet.

davinci26 replied:
fixed in 2843f25025763c74a05673d5c618d9f83ce72e1c

ggreenway commented on source/common/router/router.cc:
```diff
@@ -491,6 +493,34 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
   transport_socket_options_ = Network::TransportSocketOptionsUtility::fromFilterState(
       *callbacks_->streamInfo().filterState());
+
+  if (callbacks_->getUpstreamSocketOptions()) {
```
I'm not sure if its worth the complexity or not, but I think we can avoid an allocation in most cases:

Move this chunk under the `if(has_downstream_connection_with_redirect_records)` block below, and change `upstreamSocketOptions()` to `return (upstream_options_ != nullptr) ? upstream_options_ : callbacks_->getUpstreamSocketOptions();`

davinci26 replied:
fixed in 2843f25025763c74a05673d5c618d9f83ce72e1c

ggreenway commented on source/server/connection_handler_impl.cc:
```diff
@@ -478,6 +479,33 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
 
   auto transport_socket = filter_chain->transportSocketFactory().createTransportSocket(nullptr);
   stream_info->setDownstreamSslConnection(transport_socket->ssl());
+#ifdef WIN32
```
Now that you've added `ListenerFactoryContext` to the original_dst filter config, I think this entire block can move to that filter (which is logically where it belongs).

I think you'll need to add a method to `ListenerFilterCallbacks`, but the implementation of that interface (`ActiveTcpSocket`) already has a `stream_info_`, so I think that will be easy.

davinci26 replied:
Makes sense, I like this change it feels neater that way. Fixed in 2843f25025763c74a05673d5c618d9f83ce72e1c

ggreenway reviewed (changes requested):
A few more things to cleanup. This is getting much closer to ready though.

davinci26 commented on source/common/network/win32_redirect_records_option_impl.h:
```diff
@@ -0,0 +1,53 @@
+  bool isSupported() const;
+
+private:
+  const static envoy::config::core::v3::SocketOption::SocketState in_state_ =
```
fixed in 2843f25025763c74a05673d5c618d9f83ce72e1c

davinci26 commented on source/common/network/win32_redirect_records_option_impl.cc:
```diff
@@ -0,0 +1,55 @@
+bool Win32RedirectRecordsOptionImpl::setOption(
+    Socket& socket, envoy::config::core::v3::SocketOption::SocketState state) const {
+  if (in_state_ == state) {
+    if (!optname_.hasValue()) {
```
fixed in 2843f25025763c74a05673d5c618d9f83ce72e1c

davinci26 commented on docs/root/configuration/listeners/listener_filters/original_dst_filter.rst:
```diff
@@ -3,13 +3,42 @@
+
+Redirection is not available for use with all types of network traffic. The types of packets that are supported for redirection are shown in the following list:
+
+* TCP
```
thanks, fixed in 428068405cbb44e1b5142ffce0799dfb7d1612e9

ggreenway commented on source/common/router/router.cc:
```diff
@@ -492,6 +494,33 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
   transport_socket_options_ = Network::TransportSocketOptionsUtility::fromFilterState(
       *callbacks_->streamInfo().filterState());
+
+  auto has_downstream_connection_with_redirect_records =
```
One idea (although I'm not sure if it pans out): could `buildWFPRedirectRecordsOptions` be called when this filterState() is set, and instead of putting in a `Win32RedirectRecordsFilterState`, put in a generic list of socket options that should be applied to the upstream? `Network::UpstreamSocketOptionsFilterState` or something. It would reduce a bit of duplication between router and tcp_proxy, and would encapsulate more of the original_dst/win32 details in the original_dst filter.

davinci26 replied:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/1e5b32387ef971a8d84363bfe7ac93f01270b0db

ggreenway reviewed (changes requested):
Need to add a release note.

davinci26 commented:
@ggreenway I addressed all your comments, thanks a lot for all the reviews.

Should I drop the coverage for the `win32Ioctl` function which is not exercised in the linux code path and drops the LCOV?

ggreenway commented:
> Should I drop the coverage for the `win32Ioctl` function which is not exercised in the linux code path and drops the LCOV?

Yes, that's fine

ggreenway commented on source/common/network/win32_redirect_records_option_impl.h:
```diff
@@ -0,0 +1,41 @@
+#pragma once
```
Can/should this class/files live in the original_dst filter?

davinci26 replied:
Wouldnt that create an awkward dependency graph?

`\\source\common\network:socket_option_factory_lib` depends on `\\source\extension\filters\listener\filters\original_dst:win32_redirect_records_option_lib`

ggreenway reviewed (commented):
This is looking really good.

Is it possible to add an integration test for this, or would the tests never be run in the proper setup?

Does this change introduce any issues with matching downstream to upstream connections? My reading of the docs makes it look like it shouldn't be anything new because original_dst already maps this correctly. Would two requests for the same original_dst address have different values for the redirect-records? If that is possible, I'd like to add some tests around that. This is what the `hashKey()` on the socket option is for, and it would validate that this is working as expected.

davinci26 commented:
> Is it possible to add an integration test for this, or would the tests never be run in the proper setup?

For now the CI image does not support this feature so it is not possible. Even after that I am not sure if it would be possible to test it since everything runs inside a docker container. 

>  Would two requests for the same original_dst address have different values for the redirect-records?

yes it is possible this is why I added the hash key in the option.

in the following scenario:
1. request to google.com
2. add firewall rules to block google.com
3. the new request to google.com will have different redirect records

I will test that scenario explicitly. I was hoping that this covered by existing tests

ggreenway commented on test/common/network/win32_redirect_records_option_test.cc:
```diff
@@ -0,0 +1,88 @@
+  Win32RedirectRecordsOptionImpl socket_option{*redirect_records_};
+  socket_option.hashKey(hash);
+  std::vector<uint8_t> expected_hash;
+  pushScalarToByteVector(StringUtil::CaseInsensitiveHash()(redirect_records_data_), expected_hash);
```
I think this test could be improved by testing with two different instances, with different data for `redirect_records`, and EXPECT_NE the results of the two.

davinci26 replied:
added this test as well

davinci26 commented:
@ggreenway do we want more testing for this PR? or any other changes?

ggreenway reviewed (changes requested):
This looks good, except it needs a release-note

davinci26 commented:
I thought it was only in PR description but I guess it is also needed in `docs/root/version_history/current.rst`

pravb commented on docs/root/version_history/current.rst:
```diff
@@ -88,6 +88,7 @@ New Features
 * http: added support for `Envoy::ScopeTrackedObject` for HTTP/1 dispatching. Crashes while inside the dispatching loop should dump debug information.
 * http: added support for :ref:`preconnecting <envoy_v3_api_msg_config.cluster.v3.Cluster.PreconnectPolicy>`. Preconnecting is off by default, but recommended for clusters serving latency-sensitive traffic, especially if using HTTP/1.1.
 * http: change frame flood and abuse checks to the upstream HTTP/2 codec to ON by default. It can be disabled by setting the `envoy.reloadable_features.upstream_http2_flood_checks` runtime key to false.
+* original_dst*: added support for :ref:`Original Destination <config_listener_filters_original_dst_filter>` on Windows. With this change Envoy can be used as sidecar proxy on Windows.
```
This enables the use of Envoy as a sidecar proxy on Windows.

davinci26 replied:
fixed in c6718e872a64a57e440c090fa2df9fc6dd5a40eb

pravb commented on source/extensions/filters/listener/original_dst/config.cc:
```diff
@@ -22,10 +22,28 @@ class OriginalDstConfigFactory : public Server::Configuration::NamedListenerFilt
+    if (!Network::win32SupportsOriginalDestination()) {
+      throw EnvoyException("[Windows] Envoy was compiled without support for `SO_ORIGINAL_DST`, "
+                           "the original destination filter can not be used");
+    }
```
cant not - > cannot

davinci26 replied:
fixed in c6718e872a64a57e440c090fa2df9fc6dd5a40eb

pravb commented on source/extensions/filters/listener/original_dst/original_dst.cc:
```diff
@@ -20,12 +22,47 @@ Network::FilterStatus OriginalDstFilter::onAccept(Network::ListenerFilterCallbac
     // In this case the listener handles the connection directly and does not hand it off.
     if (original_local_address) {
+#ifdef WIN32
+      // See how to perform bind or connect redirection on MSDN
```
on MSDN -> here:

davinci26 replied:
fixed in c6718e872a64a57e440c090fa2df9fc6dd5a40eb

pravb commented on source/extensions/filters/listener/original_dst/original_dst.cc:
```diff
@@ -20,12 +22,47 @@ Network::FilterStatus OriginalDstFilter::onAccept(Network::ListenerFilterCallbac
+      // https://docs.microsoft.com/en-us/windows-hardware/drivers/network/using-bind-or-connect-redirection
+      if constexpr (Network::win32SupportsOriginalDestination()) {
+        if (traffic_direction_ == envoy::config::core::v3::OUTBOUND) {
+          ENVOY_LOG(debug, "[Windows] Querying for redirect record for outbound listener");
```
consider moving the log statement to after the Ioctl call so you can always log the returned status and maybe even the redirect record.

pravb commented on source/common/network/win32_redirect_records_option_impl.cc:
```diff
@@ -0,0 +1,77 @@
+                          const_cast<void*>(reinterpret_cast<const void*>(redirect_records_.buf_)),
+                          redirect_records_.buf_size_, nullptr, 0, &size);
+    if (result.rc_ != 0) {
+      ENVOY_LOG(warn, "Setting WFP records on socket failed: {}", errorDetails(result.errno_));
```
Consider always logging even for success case (with result.rc value) if caller doesn't.

davinci26 replied:
The caller logs if there is a failure in applying socket options and the rest of the socket options implementing this interface are only logging failures. From what I have seen the common way is to only log failures

ggreenway replied:
I think it's fine to only log failure. If non-failure logging were to be added, it should be at `trace` level.

ggreenway commented on docs/root/version_history/current.rst:
```diff
@@ -88,7 +88,7 @@ New Features
 * http: added support for :ref:`preconnecting <envoy_v3_api_msg_config.cluster.v3.Cluster.PreconnectPolicy>`. Preconnecting is off by default, but recommended for clusters serving latency-sensitive traffic, especially if using HTTP/1.1.
 * http: change frame flood and abuse checks to the upstream HTTP/2 codec to ON by default. It can be disabled by setting the `envoy.reloadable_features.upstream_http2_flood_checks` runtime key to false.
-* original_dst*: added support for :ref:`Original Destination <config_listener_filters_original_dst_filter>` on Windows. With this change Envoy can be used as sidecar proxy on Windows.
+* original_dst*: added support for :ref:`Original Destination <config_listener_filters_original_dst_filter>` on Windows. This enables the use of Envoy as a sidecar proxy on Windows.
```
`* original_dst: ....`. No `*` before `:`

ggreenway replied:
And the link target should be `<config_listener_filters_original_dst>`

davinci26 replied:
fixed, see https://storage.googleapis.com/envoy-pr/2fa1526/docs/version_history/current.html

The docs generation does not work on windows so for doc changes I usually test them via the CI

davinci26 commented:
@ggreenway is away this week, @htuch do you want to do the last review to get this PR merged?

mattklein123 commented:
I'm happy to take a look this week some time.

davinci26 commented:
Thanks Matt! This comment https://github.com/envoyproxy/envoy/pull/14753#issuecomment-763240675 gives a lot of additional context and should help a lot with the review.

htuch commented:
Do we need another review? Since Greg is senior maintainer, I think we can just go ahead and merge. But if @mattklein123 wants to take another pass that is good as well :)

htuch commented:
@davinci26 can you fix the merge conflict?

mattklein123 commented on api/envoy/api/v2/listener.proto:
```diff
@@ -195,6 +195,7 @@ message Listener {
   google.protobuf.UInt32Value tcp_fast_open_queue_length = 12;
 
   // Specifies the intended direction of the traffic relative to the local Envoy.
+  // This property is required on Windows for listeners using the original destination filter.
```
doc nit: ref link to original destination docs here and below

davinci26 replied:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/f7698962ac3e21e673937b4e3723f171b87ae8ea

mattklein123 commented on include/envoy/api/os_sys_calls.h:
```diff
@@ -185,6 +185,14 @@ class OsSysCalls {
+  /**
+   * @see MSDN WSAIoctl. Controls the mode of a socket.
+   */
+  virtual SysCallIntResult win32Ioctl(os_fd_t sockfd, unsigned long control_code, void* in_buffer,
+                                      unsigned long in_buffer_len, void* out_buffer,
+                                      unsigned long out_buffer_len,
+                                      unsigned long* bytes_returned) PURE;
```
Apologies if this was already discussed, but it seems odd to me to add this API at the global level as it is windows specific (yes I know the rest of the APIs are POSIX specific but there is some assumption that all OS support POSIX in some form). Since the syscalls are accessed by singleton, wdyt about adding a Win32 specific syscalls class that is only compiled on win32, then checked to be only called within win32 code? Since this gets called through from other APIs like socket, I'm wondering if additionally there could be a set of win32 specific free functions that know how to do various things with these special IOCTLs vs. having NOP everywhere else? The alternative would be to define some type of "extra" generic IOCTL and have it be feature based vs. OS based, but that seems like more work?

davinci26 replied:
Is the approach that you are describing to remove `win32Ioctl` from the `socket`/`io_handle` interface? And invoke this functionality through a  new  `win32OsSysCallsSingleton`?

--- 
Another approach I considered was to change the API signature of `ioctl` in `os_sys_calls.h` to mirror the Windows function signature. 

i.e.

instead of 
```
  /**
   * @see ioctl (man 2 ioctl)
   */
  virtual SysCallIntResult ioctl(os_fd_t sockfd, unsigned long int request, void* argp) PURE;
```

to have:

```
  /**
   * @see MSDN WSAIoctl. Controls the mode of a socket.
   */
  virtual SysCallIntResult win32Ioctl(os_fd_t sockfd, unsigned long control_code, void* in_buffer,
                                      unsigned long in_buffer_len, void* out_buffer,
                                      unsigned long out_buffer_len,
                                      unsigned long* bytes_returned) PURE;
```

and in POSIX the last parameters would be unused (`in_buffer` would correspond to `argp`). There are few use cases for `ioctl` in the codebase (I couldn't find any caller outside of tests).

mattklein123 replied:
I was suggesting the former, but I think the latter is fine also and simpler with appropriate comments. WDYT?

davinci26 replied:
I prefer the later, because in the former we would need to have access to the `int fd` to invoke the system calls which is not always easy since we abstract it away from the caller in the `io_socket_handle_impl`.

davinci26 replied:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/f7698962ac3e21e673937b4e3723f171b87ae8ea

mattklein123 commented on include/envoy/network/io_handle.h:
```diff
@@ -26,6 +26,21 @@ using RawSliceArrays = absl::FixedArray<absl::FixedArray<Buffer::RawSlice>>;
 
 namespace Network {
 
+struct Win32RedirectRecords {
+  // The size of the buffer is selected based on:
+  // https://docs.microsoft.com/en-us/windows-hardware/drivers/network/sio-query-wfp-connection-redirect-records
+  uint8_t buf_[1024];
+  unsigned long buf_size_;
+};
+
+constexpr bool win32SupportsOriginalDestination() {
+#if defined(WIN32) && defined(SIO_QUERY_WFP_CONNECTION_REDIRECT_RECORDS) && defined(SO_ORIGINAL_DST)
+  return true;
+#else
+  return false;
+#endif
+}
```
Per my other comment, would prefer this is in a win32 dedicated file since I think it's only called from win32 guards?

davinci26 replied:
I left it as it is since we are not adding the dedicated file.

I could move it to `platform.h` if you prefer it, open to suggestions. The reason I did not do that is because changes in `platform.h` is because it destroys my inner loop. Changing `platform.h` requires to recompile almost everything.

mattklein123 replied:
+1 to move this to platform. Understand on inner loop but once the code is finished seems better to move it there?

davinci26 replied:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/414284dd3e0cbd792817258241c60e764db48bba

mattklein123 reviewed (changes requested):
Great to have this feature parity. A few high level comments from a quick pass. Thank you!

/wait

davinci26 commented:
/retest

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -291,6 +292,11 @@ class ListenerFilterCallbacks {
+  /**
+   * @return Object on which filters can share data on a per-request basis.
+   */
+  virtual StreamInfo::FilterState& filterState() const PURE;
```
Sorry just noticed this but it seems odd that a const method returns a non-const object? Is this necessary?

davinci26 replied:
`FilterState::SetData` is not const so the return value cannot be `const`. Technically acquiring the reference is also `const` although the caller might decide to change it.

davinci26 replied:
I could make it into `virtual StreamInfo::FilterState& filterState() PURE;`

mattklein123 replied:
+1 on that change, thanks. Otherwise LGTM.

/wait

davinci26 replied:
fixed in 5fb510a5598d64837ca624a605c2af901f631ccd

mattklein123 commented on source/common/network/BUILD:
```diff
@@ -352,6 +352,23 @@ envoy_cc_library(
     ],
 )
 
+envoy_cc_library(
```
Does this only get compiled on win32? Can we make it so?

davinci26 replied:
No it is not and it was deliberate. I have no strong opinion here but here is my reasoning behind it:

Con: Increased compilation times on POSIX

Pros:  Cross platform testing

By making it into a library that gets compiled into both platforms we can have the tests running into both platforms. 

The reason I prefer that is because it makes development/refactoring easier for all developers. I want to minimize the probability that a POSIX dev want to make a refactor in this part of the code and is blocked or stuck because they can't run tests for the component.

mattklein123 replied:
OK fair enough.

mattklein123 reviewed (changes requested):
Thanks LGTM with some small remaining comments.

/wait

antoniovicente commented on test/per_file_coverage.sh:
```diff
@@ -13,7 +13,7 @@ declare -a KNOWN_LOW_COVERAGE=(
 "source/common/http/http3:50.0"
 "source/common/json:90.6"
-"source/common/network:95.2"
+"source/common/network:95.0"
```
cc @alyssawilk @junr03 

Explanations for this reduction:
- Changes in this PR
- Test coverage for connection_impl.cc went down when runtime flag in https://github.com/envoyproxy/envoy/pull/15028 was rolled back
- IoSocketHandleImpl::activateFileEvents is only covered by a death test, coverage from those is not reliable.
- RawBufferSocket::startSecureTransport not always called by tests. https://github.com/envoyproxy/envoy/pull/13112/files#r578044664

Could you add "# flaky: be careful adjusting" to this line?

davinci26 replied:
Added a comment here https://github.com/envoyproxy/envoy/pull/14753/commits/414284dd3e0cbd792817258241c60e764db48bba#diff-0ac04ef31c616358d8234bd473469c1a233ecf37947a13475cce764e7f6c43eeR16

davinci26 commented:
/retest

mattklein123 commented on include/envoy/common/platform.h:
```diff
@@ -157,7 +157,17 @@ struct msghdr {
 constexpr absl::string_view null_device_path{"NUL"};
+
+constexpr bool win32SupportsOriginalDestination() {
+#if defined(WIN32) && defined(SIO_QUERY_WFP_CONNECTION_REDIRECT_RECORDS) && defined(SO_ORIGINAL_DST)
```
This section is already guarded by MSC_VER. Do we need the WIN32 check?

davinci26 replied:
yeah left it from copy past

mattklein123 reviewed (approved):
Thanks!
